{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cfdff7dd",
   "metadata": {},
   "source": [
    "# Bespoke MCC phase skew\n",
    "\n",
    "## Use MCC peak-calls to link SNPs to CAD promoter regions\n",
    "## Use bespoke WASP input files keyed off SNP haplotypes to assess gene-skew"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "88dd5028",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "from glob import glob\n",
    "import pysam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f00c5dc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Files available on request from Matt Baxter\n",
    "matt_directory = \"<matt_directory>\"\n",
    "shared_directory = \"<shared_directory>\"\n",
    "ed_directory=\"<ed_directory>\"\n",
    "\n",
    "cad_promoter_file_path = \"/project/Wellcome_Discovery/mbaxter/00_MCC/00_Peak_calling/HUVEC_CAD_Promoters_1Mb.bed\"\n",
    "\n",
    "peak_calls_format_lookup = {\n",
    "    \"HUVEC_D1\": matt_directory + \"/00_MCC/00_Peak_calling/MCC5_HUVEC1_v2/{}_mcc_peaks.bed\",\n",
    "    \"HUVEC_D2\": matt_directory + \"/00_MCC/00_Peak_calling/MCC3_HUVEC2_v2/{}_mcc_peaks.bed\",\n",
    "    \"HUVEC_D3\": matt_directory + \"/00_MCC/00_Peak_calling/MCC6_HUVEC3_v2/{}_mcc_peaks.bed\",   \n",
    "}\n",
    "backbone_vcf_file_path = shared_directory + \"/01_in_house_data/02_HUVEC/00_WGS/00_PackBio/03_backbone/backbone.variants.vcf.gz\"\n",
    "phased_vcf_file_format = shared_directory + \"/01_in_house_data/02_HUVEC/00_WGS/00_PackBio/01_phased_VCFs_no_RefCalls/{}_phased_variants.vcf.gz\"\n",
    "bam_counts_format=ed_directory + \"/01_huvec_skew/04_bam_counts/{donor}.BAM_counts.vcf.gz\"\n",
    "point_bam_format= shared_directory + \"/01_in_house_data/02_HUVEC/03_RNASeq/01_CATCH_ALL/HUVEC_ref/results/07_indices/{}-point_hg38.bam\"\n",
    "\n",
    "\n",
    "output_test_frame_filepath = ed_directory + \"/01_huvec_skew/08_bespoke_wasp/linked.genes.tsv\"\n",
    "\n",
    "captured_huvec_enhancers_filepath = matt_directory + \"/17_CAD_paper/tables/Captured_HUVEC_enhancers_for_ED.txt\"\n",
    "dbsnp_file_path = shared_directory + \"/00_publicly_available/06_dbSNP/dbSNP.v156.hg38.vcf.gz\"\n",
    "\n",
    "wasp_input_format= ed_directory + \"/01_huvec_skew/06_wasp_files/ATAC/{donor}.WASP.input.tsv\"\n",
    "wasp_output_filepattern = ed_directory + \"/01_huvec_skew/06_wasp_files/{}/cht_results.txt\"\n",
    "\n",
    "gencode_filepath = shared_directory + \"/00_publicly_available/11_gencode_annotation/gencode.v45.primary_assembly.annotation.gtf\"\n",
    "\n",
    "bespoke_wasp_input_format = ed_directory + \"/01_huvec_skew/08_bespoke_wasp/{}.WASP.Input.2.tsv\"\n",
    "\n",
    "output_summary_filepath = ed_directory + \"/01_huvec_skew/08_bespoke_wasp/results.genes.v4.tsv\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd719b40",
   "metadata": {},
   "source": [
    "## Load in the CAD promoters and rsids of interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9c209c4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load in precomputed promoter region to gene labels\n",
    "\n",
    "cad_promoters = pd.read_csv(\n",
    "    cad_promoter_file_path, sep=\"\\t\",\n",
    "    header=None\n",
    ")[list(range(4)) + [18]]\n",
    "\n",
    "cad_promoters.columns = [\"chrom\", \"start\", \"end\", \"regmen\", \"gene_name\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6a15698b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "93\n"
     ]
    }
   ],
   "source": [
    "rsids = []\n",
    "for peak_call_format in peak_calls_format_lookup.values():\n",
    "    rsids.extend([os.path.basename(f).split(\"_\")[0] for f in glob(peak_call_format.format(\"*\"))])\n",
    "\n",
    "rsids = sorted(set(rsids))\n",
    "\n",
    "print(len(rsids))\n",
    "# print(rsids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "d7015202",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GA G\n",
      "A -\n",
      "\n",
      "C CTT\n",
      "- TT\n",
      "\n",
      "A AAGT\n",
      "- AGT\n",
      "\n",
      "Missing for rsid  chr19_11085680_C_-\n",
      "C CAG\n",
      "- AG\n",
      "\n",
      "missing, indels, n_hets, n_indel_hets, n_no_linked_genes= 3 4 37 4 32\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>TEST.SNP.POS</th>\n",
       "      <th>TEST.SNP.ID</th>\n",
       "      <th>TEST.SNP.REF.ALLELE</th>\n",
       "      <th>TEST.SNP.ALT.ALLELE</th>\n",
       "      <th>donor</th>\n",
       "      <th>TEST.SNP.GENOTYPE</th>\n",
       "      <th>TEST.SNP.HAPLOTYPE</th>\n",
       "      <th>phaseset</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2232129</td>\n",
       "      <td>rs263533</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>HUVEC_D1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0|1</td>\n",
       "      <td>587048</td>\n",
       "      <td>SKI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2232129</td>\n",
       "      <td>rs263533</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>HUVEC_D1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0|1</td>\n",
       "      <td>587048</td>\n",
       "      <td>FAAP20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2232129</td>\n",
       "      <td>rs263533</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>HUVEC_D2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1|0</td>\n",
       "      <td>588411</td>\n",
       "      <td>SKI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2232129</td>\n",
       "      <td>rs263533</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>HUVEC_D2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1|0</td>\n",
       "      <td>588411</td>\n",
       "      <td>FAAP20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2232129</td>\n",
       "      <td>rs263533</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>HUVEC_D3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1|0</td>\n",
       "      <td>589367</td>\n",
       "      <td>SKI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>chr9</td>\n",
       "      <td>129168215</td>\n",
       "      <td>rs10819474</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>HUVEC_D1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>-1</td>\n",
       "      <td>IER5L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209</th>\n",
       "      <td>chr9</td>\n",
       "      <td>129168215</td>\n",
       "      <td>rs10819474</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>HUVEC_D2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1|1</td>\n",
       "      <td>-1</td>\n",
       "      <td>NTMT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210</th>\n",
       "      <td>chr9</td>\n",
       "      <td>129168215</td>\n",
       "      <td>rs10819474</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>HUVEC_D2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1|1</td>\n",
       "      <td>-1</td>\n",
       "      <td>IER5L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>chr9</td>\n",
       "      <td>129168215</td>\n",
       "      <td>rs10819474</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>HUVEC_D3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1|0</td>\n",
       "      <td>101341845</td>\n",
       "      <td>NTMT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>chr9</td>\n",
       "      <td>129168215</td>\n",
       "      <td>rs10819474</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>HUVEC_D3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1|0</td>\n",
       "      <td>101341845</td>\n",
       "      <td>IER5L</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>213 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    CHROM  TEST.SNP.POS TEST.SNP.ID TEST.SNP.REF.ALLELE TEST.SNP.ALT.ALLELE  \\\n",
       "0    chr1       2232129    rs263533                   C                   T   \n",
       "1    chr1       2232129    rs263533                   C                   T   \n",
       "2    chr1       2232129    rs263533                   C                   T   \n",
       "3    chr1       2232129    rs263533                   C                   T   \n",
       "4    chr1       2232129    rs263533                   C                   T   \n",
       "..    ...           ...         ...                 ...                 ...   \n",
       "208  chr9     129168215  rs10819474                   T                   C   \n",
       "209  chr9     129168215  rs10819474                   T                   C   \n",
       "210  chr9     129168215  rs10819474                   T                   C   \n",
       "211  chr9     129168215  rs10819474                   T                   C   \n",
       "212  chr9     129168215  rs10819474                   T                   C   \n",
       "\n",
       "        donor  TEST.SNP.GENOTYPE TEST.SNP.HAPLOTYPE   phaseset    gene  \n",
       "0    HUVEC_D1                1.0                0|1     587048     SKI  \n",
       "1    HUVEC_D1                1.0                0|1     587048  FAAP20  \n",
       "2    HUVEC_D2                1.0                1|0     588411     SKI  \n",
       "3    HUVEC_D2                1.0                1|0     588411  FAAP20  \n",
       "4    HUVEC_D3                1.0                1|0     589367     SKI  \n",
       "..        ...                ...                ...        ...     ...  \n",
       "208  HUVEC_D1                0.0                0|0         -1   IER5L  \n",
       "209  HUVEC_D2                2.0                1|1         -1   NTMT1  \n",
       "210  HUVEC_D2                2.0                1|1         -1   IER5L  \n",
       "211  HUVEC_D3                1.0                1|0  101341845   NTMT1  \n",
       "212  HUVEC_D3                1.0                1|0  101341845   IER5L  \n",
       "\n",
       "[213 rows x 10 columns]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "backbone_vcf = pysam.VariantFile(backbone_vcf_file_path)\n",
    "\n",
    "donor_vcfs_for_phase = {\n",
    "    donor: pysam.VariantFile(phased_vcf_file_format.format(donor)) for donor in [\"HUVEC_D1\", \"HUVEC_D2\", \"HUVEC_D3\"]\n",
    "}\n",
    "\n",
    "missing = 0\n",
    "indels = 0\n",
    "n_hets = 0\n",
    "n_indel_hets = 0\n",
    "n_no_linked_genes = 0\n",
    "\n",
    "\n",
    "test_frame = []\n",
    "for _, row in rsid_definitions.iterrows():\n",
    "    variants = list(backbone_vcf.fetch(row.chrom, row.pos - 1, row.pos))\n",
    "    \n",
    "    peaks_lookup = {\n",
    "        x: pd.read_csv(y.format(row.rsid), sep=\"\\t\") for x, y in peak_calls_format_lookup.items() if \n",
    "        os.path.exists(y.format(row.rsid))\n",
    "    }\n",
    "    if len(peaks_lookup) == 0:\n",
    "        print(\"Missing for rsid \", row.rsid)\n",
    "        continue\n",
    "\n",
    "    # Use peak files to link SNPs to genes via CAD promoters\n",
    "    linked_genes = set()\n",
    "    for donor, peaks in peaks_lookup.items():\n",
    "        for _, peak in peaks.iterrows():            \n",
    "            cad_genes = cad_promoters[\n",
    "                (cad_promoters.chrom == peak.Chromosome) &\n",
    "                (cad_promoters.start <= peak.End + 100) &\n",
    "                (cad_promoters.end >= peak.Start - 100)\n",
    "            ]\n",
    "            for gene_name in set(cad_genes.gene_name):\n",
    "                linked_genes.add(gene_name)\n",
    "    \n",
    "    linked_genes = set(linked_genes).intersection(set(genes.gene_name))\n",
    "    if len(linked_genes) == 0:\n",
    "        n_no_linked_genes += 1\n",
    "        continue\n",
    "\n",
    "\n",
    "    # Build up the annotations to link variants to genes. Use WASP-style column labels\n",
    "    if len(variants) == 1:\n",
    "        assert variants[0].pos == row.pos\n",
    "        if len(variants[0].ref) == len(variants[0].alts[0]):\n",
    "            assert row.ref in {variants[0].ref, variants[0].alts[0]}\n",
    "            assert row.alt in {variants[0].ref, variants[0].alts[0]}\n",
    "        else:\n",
    "            print(variants[0].ref, variants[0].alts[0])\n",
    "            print(row.ref, row.alt)\n",
    "            indels += 1\n",
    "            if any((set(var.samples[don][\"GT\"]) == {0, 1} for don in donors)):\n",
    "                n_indel_hets += 1\n",
    "            print()\n",
    "        var = variants[0]\n",
    "\n",
    "        donors = [\n",
    "            \"HUVEC_D1\", \"HUVEC_D2\", \"HUVEC_D3\",\n",
    "        ]\n",
    "        if any((set(var.samples[don][\"GT\"]) == {0, 1} for don in donors)):\n",
    "            snp_label = {\n",
    "               \"CHROM\": var.chrom,\n",
    "                \"TEST.SNP.POS\": var.pos,\n",
    "                \"TEST.SNP.ID\": var.id,\n",
    "                \"TEST.SNP.REF.ALLELE\": var.ref,\n",
    "                \"TEST.SNP.ALT.ALLELE\": var.alts[0],\n",
    "            }\n",
    "            for donor in donors:\n",
    "                snp_label_donor = snp_label.copy()\n",
    "                snp_label_donor[\"donor\"] = donor\n",
    "\n",
    "                if set(var.samples[donor][\"GT\"]) == {None}:\n",
    "                    snp_label_donor[\"TEST.SNP.GENOTYPE\"] = 0.0\n",
    "                    snp_label_donor[\"TEST.SNP.HAPLOTYPE\"] = \"0|0\"\n",
    "                    snp_label_donor[\"phaseset\"] = -1\n",
    "                elif set(var.samples[donor][\"GT\"]) == {0, 1}:\n",
    "                    snp_label_donor[\"TEST.SNP.GENOTYPE\"] = 1.0\n",
    "                    snp_label_donor[\"TEST.SNP.HAPLOTYPE\"] = \"|\".join(\n",
    "                        list(map(str, var.samples[donor][\"GT\"]))\n",
    "                    )\n",
    "                    \n",
    "                    phase_vars = list(donor_vcfs_for_phase[donor].fetch(var.chrom, var.pos-1, var.pos))\n",
    "                    assert len(phase_vars) == 1\n",
    "                    phase_var = phase_vars[0]\n",
    "                    assert phase_var.ref == var.ref and phase_var.alts == var.alts\n",
    "                    \n",
    "                    snp_label_donor[\"phaseset\"] = phase_var.samples[donor].get(\"PS\", -1)\n",
    "                else:\n",
    "                    assert set(var.samples[donor][\"GT\"]) == {1}\n",
    "                    snp_label_donor[\"TEST.SNP.GENOTYPE\"] = 2.0\n",
    "                    snp_label_donor[\"TEST.SNP.HAPLOTYPE\"] = \"1|1\"\n",
    "                    snp_label_donor[\"phaseset\"] = -1\n",
    "                \n",
    "                for gene in linked_genes:\n",
    "                    snp_label_donor_gene = snp_label_donor.copy()\n",
    "                    snp_label_donor_gene[\"gene\"] = gene\n",
    "                    test_frame.append(snp_label_donor_gene)\n",
    "            \n",
    "            n_hets += 1\n",
    "            \n",
    "            \n",
    "        \n",
    "    elif len(variants) == 0:\n",
    "        missing += 1\n",
    "    else:\n",
    "        print(\"Have 2 here\")\n",
    "        pass\n",
    "    \n",
    "print(\n",
    "    \"missing, indels, n_hets, n_indel_hets, n_no_linked_genes=\",\n",
    "    missing, indels, n_hets, n_indel_hets, n_no_linked_genes\n",
    ")\n",
    "test_frame = pd.DataFrame(test_frame)\n",
    "test_frame.to_csv(output_test_frame_filepath, sep=\"\\t\", index=False)\n",
    "test_frame\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "9e5d1507",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Use dbSNP to remap rsID labels\n",
    "\n",
    "rsid_definitions = pd.read_csv(captured_huvec_enhancers_filepath, sep=\"\\t\", header=None)\n",
    "\n",
    "rsid_definitions.columns = [\"chrom\", \"pos\", \"ref\", \"alt\", \"rsid\"]\n",
    "db_snp = pysam.VariantFile(db_snp_filepath)\n",
    "rsid_mapping = {}\n",
    "for _, row in rsid_definitions.iterrows():\n",
    "    db_def = [x for x in db_snp.fetch(row.chrom, row.pos-2, row.pos+2)]\n",
    "    if row.rsid not in {x.id for x in db_def}:\n",
    "        new_records = [x for x in db_def if x.pos == row.pos]\n",
    "        if len(new_records) != 1:\n",
    "            continue\n",
    "        new_record = new_records[0]\n",
    "        \n",
    "        rsids.append(new_record.id)\n",
    "        rsid_mapping[new_record.id] = row.rsid\n",
    "\n",
    "rsids = sorted(set(rsids))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "aed689d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load in precomputed WASP output. This is the ATAC skew not central to CAD paper\n",
    "\n",
    "common_cols = 'TEST.SNP.CHROM', 'TEST.SNP.POS', 'TEST.SNP.ID', 'TEST.SNP.REF.ALLELE','TEST.SNP.ALT.ALLELE', 'REGION.START', 'REGION.END'\n",
    "\n",
    "data_type=\"ATAC\"\n",
    "wasp_output=pd.read_csv(wasp_output_filepattern.format(data_type), sep=\"\\t\")\n",
    "\n",
    "other_cols = set(wasp_output.columns) - set(common_cols)\n",
    "rename = {x: \"{}.{}\".format(x, data_type) for x in other_cols}\n",
    "\n",
    "wasp_output = wasp_output.rename(columns=rename)\n",
    "wasp_output[\"SKEW.{}\".format(data_type)] = wasp_output[\"REF.AS.READ.COUNT.{}\".format(data_type)] / wasp_output[\"TOTAL.AS.READ.COUNT.{}\".format(data_type)]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63dde184",
   "metadata": {},
   "source": [
    "# Use gene-code gene annotations to define POINT-seq regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a275f32f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>chr1</td>\n",
       "      <td>65419</td>\n",
       "      <td>71585</td>\n",
       "      <td>+</td>\n",
       "      <td>OR4F5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>314</th>\n",
       "      <td>chr1</td>\n",
       "      <td>450740</td>\n",
       "      <td>451678</td>\n",
       "      <td>-</td>\n",
       "      <td>OR4F29</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom   start     end strand gene_name\n",
       "58   chr1   65419   71585      +     OR4F5\n",
       "314  chr1  450740  451678      -    OR4F29"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load in gene database\n",
    "\n",
    "genes = pd.read_csv(gencode_filepath, skiprows=5, header=None, sep=\"\\t\")\n",
    "genes = genes[genes[2] == \"gene\"]\n",
    "genes = genes[genes[8].str.contains(\"protein\")]\n",
    "genes = genes[genes[0].str.startswith(\"chr\")]\n",
    "\n",
    "genes[9] = genes[8].str.split(\";\").apply(\n",
    "    lambda x: [y for y in x if \"gene_name\" in y][0]\n",
    ").str.split(\"gene_name\").apply(lambda x: x[1]).str.strip().str.strip(\"\\\"\")\n",
    "\n",
    "genes = genes[[0, 3, 4, 6, 9]]\n",
    "\n",
    "genes.columns = [\"chrom\", \"start\", \"end\", \"strand\", \"gene_name\"]\n",
    "\n",
    "genes.head(n=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ebb54278",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HUVEC_D1\n",
      "HUVEC_D2\n",
      "HUVEC_D3\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>allele1</th>\n",
       "      <th>allele2</th>\n",
       "      <th>rna_seq_allele1</th>\n",
       "      <th>rna_seq_allele2</th>\n",
       "      <th>donor</th>\n",
       "      <th>phaseset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14630</th>\n",
       "      <td>MMP2</td>\n",
       "      <td>137</td>\n",
       "      <td>140</td>\n",
       "      <td>766</td>\n",
       "      <td>771</td>\n",
       "      <td>HUVEC_D1</td>\n",
       "      <td>54252436.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34679</th>\n",
       "      <td>MMP2</td>\n",
       "      <td>1183</td>\n",
       "      <td>1349</td>\n",
       "      <td>611</td>\n",
       "      <td>637</td>\n",
       "      <td>HUVEC_D2</td>\n",
       "      <td>46382627.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54728</th>\n",
       "      <td>MMP2</td>\n",
       "      <td>546</td>\n",
       "      <td>535</td>\n",
       "      <td>331</td>\n",
       "      <td>342</td>\n",
       "      <td>HUVEC_D3</td>\n",
       "      <td>46382645.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene  allele1  allele2  rna_seq_allele1  rna_seq_allele2     donor  \\\n",
       "14630  MMP2      137      140              766              771  HUVEC_D1   \n",
       "34679  MMP2     1183     1349              611              637  HUVEC_D2   \n",
       "54728  MMP2      546      535              331              342  HUVEC_D3   \n",
       "\n",
       "         phaseset  \n",
       "14630  54252436.0  \n",
       "34679  46382627.0  \n",
       "54728  46382645.0  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# For all donors form grid of point values\n",
    "# One VCf is only used for the phase count which isn't present for allele counts.\n",
    "\n",
    "donors = [\"HUVEC_D1\", \"HUVEC_D2\", \"HUVEC_D3\"]\n",
    "\n",
    "data = []\n",
    "\n",
    "for donor in donors:\n",
    "    print(donor)\n",
    "    \n",
    "    vcf = pysam.VariantFile(phased_vcf_file_format.format(donor))\n",
    "    vcf_counts = pysam.VariantFile(bam_counts_format.format(donor=donor))\n",
    "\n",
    "    for _, gene in genes.iterrows():\n",
    "        strand = gene.strand == \"+\"\n",
    "\n",
    "        vcf_variants = list(vcf.fetch(gene.chrom, gene.start, gene.end))\n",
    "        vcf_variants_counts = list(vcf_counts.fetch(gene.chrom, gene.start, gene.end))\n",
    "\n",
    "        phase = None\n",
    "        point_allele1 = 0\n",
    "        point_allele2 = 0\n",
    "        \n",
    "        rna_seq_allele1 = 0\n",
    "        rna_seq_allele2 = 0\n",
    "        \n",
    "        for var, var_counts in zip(vcf_variants, vcf_variants_counts):\n",
    "            if len(var.ref) > 1:\n",
    "                continue\n",
    "            if len(var.alts) > 1:\n",
    "                continue\n",
    "            if len(var.alts[0]) > 1:\n",
    "                continue\n",
    "            \n",
    "            new_phase = var.samples[donor].get(\"PS\")\n",
    "            if new_phase is None:\n",
    "                continue\n",
    "\n",
    "            if phase is None:\n",
    "                phase = new_phase\n",
    "            elif phase != new_phase:\n",
    "                break\n",
    "            \n",
    "            assert var.id == var_counts.id\n",
    "            assert var.samples[donor][\"GT\"] == var_counts.samples[donor][\"GT\"]\n",
    "\n",
    "            allele_code1 = \"R\" if var.samples[donor][\"GT\"] == (0, 1) else \"A\"\n",
    "            allele_code2 = \"A\" if var.samples[donor][\"GT\"] == (0, 1) else \"R\"\n",
    "            strand_code = \"F\" if strand else \"R\"\n",
    "            \n",
    "            point_allele1 += var_counts.info.get(\"P\" + strand_code + allele_code1, 0)\n",
    "            point_allele2 += var_counts.info.get(\"P\" + strand_code + allele_code2, 0)\n",
    "            \n",
    "            rna_seq_allele1 += var_counts.info.get(\"R\" + strand_code + allele_code1, 0)\n",
    "            rna_seq_allele2 += var_counts.info.get(\"R\" + strand_code + allele_code2, 0)\n",
    "        \n",
    "\n",
    "        data.append({\n",
    "            \"gene\": gene.gene_name,\n",
    "            \"allele1\": point_allele1,\n",
    "            \"allele2\": point_allele2,\n",
    "            \"rna_seq_allele1\": rna_seq_allele1,\n",
    "            \"rna_seq_allele2\": rna_seq_allele2,\n",
    "            \"donor\": donor,\n",
    "            \"phaseset\": phase,\n",
    "        })\n",
    "    \n",
    "point = pd.DataFrame(data)\n",
    "\n",
    "point[point.gene == \"MMP2\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "b752b7fb",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>TEST.SNP.POS</th>\n",
       "      <th>TEST.SNP.ID</th>\n",
       "      <th>TEST.SNP.REF.ALLELE</th>\n",
       "      <th>TEST.SNP.ALT.ALLELE</th>\n",
       "      <th>TEST.SNP.GENOTYPE</th>\n",
       "      <th>TEST.SNP.HAPLOTYPE</th>\n",
       "      <th>REGION.START</th>\n",
       "      <th>REGION.END</th>\n",
       "      <th>REGION.SNP.POS</th>\n",
       "      <th>REGION.SNP.HET.PROB</th>\n",
       "      <th>REGION.SNP.LINKAGE.PROB</th>\n",
       "      <th>REGION.SNP.REF.HAP.COUNT</th>\n",
       "      <th>REGION.SNP.ALT.HAP.COUNT</th>\n",
       "      <th>REGION.SNP.OTHER.HAP.COUNT</th>\n",
       "      <th>REGION.READ.COUNT</th>\n",
       "      <th>GENOMEWIDE.READ.COUNT</th>\n",
       "      <th>donor</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2232129</td>\n",
       "      <td>rs263533</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0|1</td>\n",
       "      <td>2231150</td>\n",
       "      <td>2233061</td>\n",
       "      <td>2231505;2232129;2232677</td>\n",
       "      <td>0.9815737349020663;0.9860587390115204;0.986721...</td>\n",
       "      <td>0.5;0.5;0.5</td>\n",
       "      <td>4;7;1</td>\n",
       "      <td>1;7;3</td>\n",
       "      <td>0;0;0</td>\n",
       "      <td>374</td>\n",
       "      <td>101614726</td>\n",
       "      <td>HUVEC_D1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2299</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24705677</td>\n",
       "      <td>rs66530629</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0|0</td>\n",
       "      <td>24704974</td>\n",
       "      <td>24706046</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>495</td>\n",
       "      <td>101614726</td>\n",
       "      <td>HUVEC_D1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2305</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24724734</td>\n",
       "      <td>rs1316829</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0|0</td>\n",
       "      <td>24724401</td>\n",
       "      <td>24725857</td>\n",
       "      <td>24724662</td>\n",
       "      <td>0.4891842560813785</td>\n",
       "      <td>0.5</td>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>467</td>\n",
       "      <td>101614726</td>\n",
       "      <td>HUVEC_D1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4033</th>\n",
       "      <td>chr1</td>\n",
       "      <td>56411837</td>\n",
       "      <td>rs11206803</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0|0</td>\n",
       "      <td>56411187</td>\n",
       "      <td>56412269</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>407</td>\n",
       "      <td>101614726</td>\n",
       "      <td>HUVEC_D1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8518</th>\n",
       "      <td>chr1</td>\n",
       "      <td>175169466</td>\n",
       "      <td>rs10912901</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0|0</td>\n",
       "      <td>175168803</td>\n",
       "      <td>175170501</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1118</td>\n",
       "      <td>101614726</td>\n",
       "      <td>HUVEC_D1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119272</th>\n",
       "      <td>chr8</td>\n",
       "      <td>133546137</td>\n",
       "      <td>rs16904940</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0|0</td>\n",
       "      <td>133545478</td>\n",
       "      <td>133546745</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>760</td>\n",
       "      <td>227790954</td>\n",
       "      <td>HUVEC_D3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121120</th>\n",
       "      <td>chr9</td>\n",
       "      <td>22103342</td>\n",
       "      <td>rs1537373</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0|0</td>\n",
       "      <td>22103207</td>\n",
       "      <td>22103972</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>479</td>\n",
       "      <td>227790954</td>\n",
       "      <td>HUVEC_D3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123971</th>\n",
       "      <td>chr9</td>\n",
       "      <td>107755513</td>\n",
       "      <td>rs944172</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0|0</td>\n",
       "      <td>107754953</td>\n",
       "      <td>107756029</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>649</td>\n",
       "      <td>227790954</td>\n",
       "      <td>HUVEC_D3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125157</th>\n",
       "      <td>chr9</td>\n",
       "      <td>127845239</td>\n",
       "      <td>rs11789185</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>2.00</td>\n",
       "      <td>1|1</td>\n",
       "      <td>127844614</td>\n",
       "      <td>127845884</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>914</td>\n",
       "      <td>227790954</td>\n",
       "      <td>HUVEC_D3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125337</th>\n",
       "      <td>chr9</td>\n",
       "      <td>129168215</td>\n",
       "      <td>rs10819474</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1|0</td>\n",
       "      <td>129166780</td>\n",
       "      <td>129168455</td>\n",
       "      <td>129167678;129168215</td>\n",
       "      <td>0.989924900833265;0.9898863328314084</td>\n",
       "      <td>0.5;0.5</td>\n",
       "      <td>3;25</td>\n",
       "      <td>3;17</td>\n",
       "      <td>0;0</td>\n",
       "      <td>885</td>\n",
       "      <td>227790954</td>\n",
       "      <td>HUVEC_D3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>228 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       CHROM  TEST.SNP.POS TEST.SNP.ID TEST.SNP.REF.ALLELE  \\\n",
       "235     chr1       2232129    rs263533                   C   \n",
       "2299    chr1      24705677  rs66530629                   G   \n",
       "2305    chr1      24724734   rs1316829                   C   \n",
       "4033    chr1      56411837  rs11206803                   C   \n",
       "8518    chr1     175169466  rs10912901                   G   \n",
       "...      ...           ...         ...                 ...   \n",
       "119272  chr8     133546137  rs16904940                   C   \n",
       "121120  chr9      22103342   rs1537373                   T   \n",
       "123971  chr9     107755513    rs944172                   C   \n",
       "125157  chr9     127845239  rs11789185                   T   \n",
       "125337  chr9     129168215  rs10819474                   T   \n",
       "\n",
       "       TEST.SNP.ALT.ALLELE  TEST.SNP.GENOTYPE TEST.SNP.HAPLOTYPE  \\\n",
       "235                      T               1.00                0|1   \n",
       "2299                     A               0.01                0|0   \n",
       "2305                     T               0.01                0|0   \n",
       "4033                     T               0.01                0|0   \n",
       "8518                     A               0.01                0|0   \n",
       "...                    ...                ...                ...   \n",
       "119272                   T               0.01                0|0   \n",
       "121120                   G               0.01                0|0   \n",
       "123971                   T               0.01                0|0   \n",
       "125157                   G               2.00                1|1   \n",
       "125337                   C               1.00                1|0   \n",
       "\n",
       "        REGION.START  REGION.END           REGION.SNP.POS  \\\n",
       "235          2231150     2233061  2231505;2232129;2232677   \n",
       "2299        24704974    24706046                      NaN   \n",
       "2305        24724401    24725857                 24724662   \n",
       "4033        56411187    56412269                      NaN   \n",
       "8518       175168803   175170501                      NaN   \n",
       "...              ...         ...                      ...   \n",
       "119272     133545478   133546745                      NaN   \n",
       "121120      22103207    22103972                      NaN   \n",
       "123971     107754953   107756029                      NaN   \n",
       "125157     127844614   127845884                      NaN   \n",
       "125337     129166780   129168455      129167678;129168215   \n",
       "\n",
       "                                      REGION.SNP.HET.PROB  \\\n",
       "235     0.9815737349020663;0.9860587390115204;0.986721...   \n",
       "2299                                                  NaN   \n",
       "2305                                   0.4891842560813785   \n",
       "4033                                                  NaN   \n",
       "8518                                                  NaN   \n",
       "...                                                   ...   \n",
       "119272                                                NaN   \n",
       "121120                                                NaN   \n",
       "123971                                                NaN   \n",
       "125157                                                NaN   \n",
       "125337               0.989924900833265;0.9898863328314084   \n",
       "\n",
       "       REGION.SNP.LINKAGE.PROB REGION.SNP.REF.HAP.COUNT  \\\n",
       "235                0.5;0.5;0.5                    4;7;1   \n",
       "2299                       NaN                      NaN   \n",
       "2305                       0.5                        6   \n",
       "4033                       NaN                      NaN   \n",
       "8518                       NaN                      NaN   \n",
       "...                        ...                      ...   \n",
       "119272                     NaN                      NaN   \n",
       "121120                     NaN                      NaN   \n",
       "123971                     NaN                      NaN   \n",
       "125157                     NaN                      NaN   \n",
       "125337                 0.5;0.5                     3;25   \n",
       "\n",
       "       REGION.SNP.ALT.HAP.COUNT REGION.SNP.OTHER.HAP.COUNT  REGION.READ.COUNT  \\\n",
       "235                       1;7;3                      0;0;0                374   \n",
       "2299                        NaN                        NaN                495   \n",
       "2305                         10                          0                467   \n",
       "4033                        NaN                        NaN                407   \n",
       "8518                        NaN                        NaN               1118   \n",
       "...                         ...                        ...                ...   \n",
       "119272                      NaN                        NaN                760   \n",
       "121120                      NaN                        NaN                479   \n",
       "123971                      NaN                        NaN                649   \n",
       "125157                      NaN                        NaN                914   \n",
       "125337                     3;17                        0;0                885   \n",
       "\n",
       "        GENOMEWIDE.READ.COUNT     donor  \n",
       "235                 101614726  HUVEC_D1  \n",
       "2299                101614726  HUVEC_D1  \n",
       "2305                101614726  HUVEC_D1  \n",
       "4033                101614726  HUVEC_D1  \n",
       "8518                101614726  HUVEC_D1  \n",
       "...                       ...       ...  \n",
       "119272              227790954  HUVEC_D3  \n",
       "121120              227790954  HUVEC_D3  \n",
       "123971              227790954  HUVEC_D3  \n",
       "125157              227790954  HUVEC_D3  \n",
       "125337              227790954  HUVEC_D3  \n",
       "\n",
       "[228 rows x 18 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Load in the pre-computed wasp input files for the 3 donors\n",
    "\n",
    "donors = [\"HUVEC_D{}\".format(i) for i in range(1, 4)]\n",
    "\n",
    "total = []\n",
    "for donor in donors:\n",
    "    wasp_input_files = pd.read_csv(wasp_input_format.format(donor=donor), sep=\" \")\n",
    "    wasp_input_files = wasp_input_files[wasp_input_files[\"TEST.SNP.ID\"].isin(rsids)]\n",
    "    wasp_input_files[\"donor\"] = donor\n",
    "    total.append(wasp_input_files)\n",
    "\n",
    "wasp_input_files = pd.concat(total)\n",
    "display(wasp_input_files)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ba8430c",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "## Create the WASP input files\n",
    "\n",
    "These each link activity rsid (in enhancer/promoter peak) to haplotype counts along gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "599940ff",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "428586390\n",
      "1|0\n",
      "1|0\n",
      "1|1\n",
      "0|1\n",
      "0|1\n",
      "0|1\n",
      "1|0\n",
      "1|0\n",
      "1|0\n",
      "0|0\n",
      "0|1\n",
      "0|0\n",
      "1|0\n",
      "1|0\n",
      "1|0\n",
      "1|0\n",
      "1|0\n",
      "1|0\n",
      "1|0\n",
      "1|0\n",
      "0|1\n",
      "0|1\n",
      "0|1\n",
      "0|1\n",
      "0|0\n",
      "1|0\n",
      "1|0\n",
      "1|0\n",
      "1|1\n",
      "1|1\n",
      "1|1\n",
      "0|0\n",
      "1|0\n",
      "1|0\n",
      "1|1\n",
      "1|1\n",
      "0|0\n",
      "0|0\n",
      "0|0\n",
      "1|0\n",
      "0|0\n",
      "1|0\n",
      "1|0\n",
      "1|0\n",
      "1|0\n",
      "1|0\n",
      "1|0\n",
      "1|0\n",
      "1|1\n",
      "1|1\n",
      "1|0\n",
      "1|1\n",
      "1|1\n",
      "0|0\n",
      "1|0\n",
      "1|0\n",
      "0|0\n",
      "0|1\n",
      "0|1\n",
      "0|1\n",
      "0|0\n",
      "0|0\n",
      "1|0\n",
      "0|0\n",
      "0|0\n",
      "0|0\n",
      "0|0\n",
      "0|0\n",
      "0|0\n",
      "1|0\n",
      "1|0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>TEST.SNP.POS</th>\n",
       "      <th>TEST.SNP.ID</th>\n",
       "      <th>TEST.SNP.REF.ALLELE</th>\n",
       "      <th>TEST.SNP.ALT.ALLELE</th>\n",
       "      <th>TEST.SNP.GENOTYPE</th>\n",
       "      <th>TEST.SNP.HAPLOTYPE</th>\n",
       "      <th>REGION.START</th>\n",
       "      <th>REGION.END</th>\n",
       "      <th>REGION.SNP.POS</th>\n",
       "      <th>REGION.SNP.HET.PROB</th>\n",
       "      <th>REGION.SNP.LINKAGE.PROB</th>\n",
       "      <th>REGION.SNP.REF.HAP.COUNT</th>\n",
       "      <th>REGION.SNP.ALT.HAP.COUNT</th>\n",
       "      <th>REGION.SNP.OTHER.HAP.COUNT</th>\n",
       "      <th>REGION.READ.COUNT</th>\n",
       "      <th>GENOMEWIDE.READ.COUNT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2232129</td>\n",
       "      <td>rs263533</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1|0</td>\n",
       "      <td>2227388</td>\n",
       "      <td>2310213</td>\n",
       "      <td>2230051;2231034;2231505;2232129;2232677;223326...</td>\n",
       "      <td>1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1....</td>\n",
       "      <td>1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1....</td>\n",
       "      <td>111;194;217;118;148;6;150;198;78;121;160;163;2...</td>\n",
       "      <td>130;194;145;135;167;3;120;170;61;112;126;143;1...</td>\n",
       "      <td>1;0;4;0;0;0;1;0;0;0;0;0;0;0;0;0;0;0;0;1;0;1;0;...</td>\n",
       "      <td>300944</td>\n",
       "      <td>428586390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2232129</td>\n",
       "      <td>rs263533</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1|0</td>\n",
       "      <td>2184461</td>\n",
       "      <td>2212720</td>\n",
       "      <td>2185716;2186900;2186920;2194537;2195995;220161...</td>\n",
       "      <td>1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1....</td>\n",
       "      <td>1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1....</td>\n",
       "      <td>4;2;1;11;4;0;0;0;3;3;3;0;3;1;3;2;4;0;2</td>\n",
       "      <td>9;6;5;11;1;0;0;0;2;2;1;0;4;1;1;1;2;0;4</td>\n",
       "      <td>0;0;0;3;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0</td>\n",
       "      <td>11727</td>\n",
       "      <td>428586390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>56411837</td>\n",
       "      <td>rs11206803</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1|1</td>\n",
       "      <td>56494761</td>\n",
       "      <td>56645301</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>33410</td>\n",
       "      <td>428586390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>205141789</td>\n",
       "      <td>rs3902193</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0|1</td>\n",
       "      <td>205227946</td>\n",
       "      <td>205285632</td>\n",
       "      <td>205230014;205230312;205231592;205232197;205234...</td>\n",
       "      <td>1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1....</td>\n",
       "      <td>1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1....</td>\n",
       "      <td>4;1;0;0;2;2;1;1;2;0;1;1;1;3;1;6;6;0;1;4;0;7;2;...</td>\n",
       "      <td>2;1;2;1;2;2;1;0;2;2;2;1;1;1;1;8;2;1;1;2;2;5;2;...</td>\n",
       "      <td>0;0;0;0;0;0;0;0;0;0;0;0;0;1;0;0;0;0;0;0;0;0;0;...</td>\n",
       "      <td>3039</td>\n",
       "      <td>428586390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>205141789</td>\n",
       "      <td>rs3902193</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0|1</td>\n",
       "      <td>205142505</td>\n",
       "      <td>205211702</td>\n",
       "      <td>205142633;205145537;205145971;205149059;205149...</td>\n",
       "      <td>1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1....</td>\n",
       "      <td>1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1....</td>\n",
       "      <td>5;2;3;5;3;2;2;11;2;5;3;11;18;1;1;0;3;0;4;8;9;1...</td>\n",
       "      <td>2;3;7;7;4;2;1;13;2;5;3;4;16;4;1;0;6;1;4;6;9;11...</td>\n",
       "      <td>0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;1;0;0;...</td>\n",
       "      <td>17576</td>\n",
       "      <td>428586390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>chr6</td>\n",
       "      <td>108902728</td>\n",
       "      <td>rs11961585</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>108848416</td>\n",
       "      <td>108974476</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>6670</td>\n",
       "      <td>428586390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>chr8</td>\n",
       "      <td>133540375</td>\n",
       "      <td>rs59413766</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>133454848</td>\n",
       "      <td>133571926</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>169811</td>\n",
       "      <td>428586390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>chr8</td>\n",
       "      <td>133546137</td>\n",
       "      <td>rs16904940</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>133454848</td>\n",
       "      <td>133571926</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>169811</td>\n",
       "      <td>428586390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>chr9</td>\n",
       "      <td>129168215</td>\n",
       "      <td>rs10819474</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1|0</td>\n",
       "      <td>129608884</td>\n",
       "      <td>129636135</td>\n",
       "      <td>129608952;129611051;129611734;129611887;129612...</td>\n",
       "      <td>1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1....</td>\n",
       "      <td>1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1....</td>\n",
       "      <td>1;0;0;0;1;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;1;0;...</td>\n",
       "      <td>0;0;1;1;0;0;0;1;1;0;0;1;1;0;0;0;0;0;0;0;0;0;0;...</td>\n",
       "      <td>0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;...</td>\n",
       "      <td>7009</td>\n",
       "      <td>428586390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>chr9</td>\n",
       "      <td>129168215</td>\n",
       "      <td>rs10819474</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1|0</td>\n",
       "      <td>129175552</td>\n",
       "      <td>129178261</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>769</td>\n",
       "      <td>428586390</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>71 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM  TEST.SNP.POS TEST.SNP.ID TEST.SNP.REF.ALLELE TEST.SNP.ALT.ALLELE  \\\n",
       "0   chr1       2232129    rs263533                   C                   T   \n",
       "1   chr1       2232129    rs263533                   C                   T   \n",
       "2   chr1      56411837  rs11206803                   C                   T   \n",
       "3   chr1     205141789   rs3902193                   C                   A   \n",
       "4   chr1     205141789   rs3902193                   C                   A   \n",
       "..   ...           ...         ...                 ...                 ...   \n",
       "66  chr6     108902728  rs11961585                   G                   A   \n",
       "67  chr8     133540375  rs59413766                   C                   A   \n",
       "68  chr8     133546137  rs16904940                   C                   T   \n",
       "69  chr9     129168215  rs10819474                   T                   C   \n",
       "70  chr9     129168215  rs10819474                   T                   C   \n",
       "\n",
       "    TEST.SNP.GENOTYPE TEST.SNP.HAPLOTYPE  REGION.START  REGION.END  \\\n",
       "0                 1.0                1|0       2227388     2310213   \n",
       "1                 1.0                1|0       2184461     2212720   \n",
       "2                 2.0                1|1      56494761    56645301   \n",
       "3                 1.0                0|1     205227946   205285632   \n",
       "4                 1.0                0|1     205142505   205211702   \n",
       "..                ...                ...           ...         ...   \n",
       "66                0.0                0|0     108848416   108974476   \n",
       "67                0.0                0|0     133454848   133571926   \n",
       "68                0.0                0|0     133454848   133571926   \n",
       "69                1.0                1|0     129608884   129636135   \n",
       "70                1.0                1|0     129175552   129178261   \n",
       "\n",
       "                                       REGION.SNP.POS  \\\n",
       "0   2230051;2231034;2231505;2232129;2232677;223326...   \n",
       "1   2185716;2186900;2186920;2194537;2195995;220161...   \n",
       "2                                                  NA   \n",
       "3   205230014;205230312;205231592;205232197;205234...   \n",
       "4   205142633;205145537;205145971;205149059;205149...   \n",
       "..                                                ...   \n",
       "66                                                 NA   \n",
       "67                                                 NA   \n",
       "68                                                 NA   \n",
       "69  129608952;129611051;129611734;129611887;129612...   \n",
       "70                                                 NA   \n",
       "\n",
       "                                  REGION.SNP.HET.PROB  \\\n",
       "0   1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1....   \n",
       "1   1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1....   \n",
       "2                                                  NA   \n",
       "3   1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1....   \n",
       "4   1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1....   \n",
       "..                                                ...   \n",
       "66                                                 NA   \n",
       "67                                                 NA   \n",
       "68                                                 NA   \n",
       "69  1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1....   \n",
       "70                                                 NA   \n",
       "\n",
       "                              REGION.SNP.LINKAGE.PROB  \\\n",
       "0   1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1....   \n",
       "1   1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1....   \n",
       "2                                                  NA   \n",
       "3   1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1....   \n",
       "4   1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1....   \n",
       "..                                                ...   \n",
       "66                                                 NA   \n",
       "67                                                 NA   \n",
       "68                                                 NA   \n",
       "69  1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1.0;1....   \n",
       "70                                                 NA   \n",
       "\n",
       "                             REGION.SNP.REF.HAP.COUNT  \\\n",
       "0   111;194;217;118;148;6;150;198;78;121;160;163;2...   \n",
       "1              4;2;1;11;4;0;0;0;3;3;3;0;3;1;3;2;4;0;2   \n",
       "2                                                  NA   \n",
       "3   4;1;0;0;2;2;1;1;2;0;1;1;1;3;1;6;6;0;1;4;0;7;2;...   \n",
       "4   5;2;3;5;3;2;2;11;2;5;3;11;18;1;1;0;3;0;4;8;9;1...   \n",
       "..                                                ...   \n",
       "66                                                 NA   \n",
       "67                                                 NA   \n",
       "68                                                 NA   \n",
       "69  1;0;0;0;1;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;1;0;...   \n",
       "70                                                 NA   \n",
       "\n",
       "                             REGION.SNP.ALT.HAP.COUNT  \\\n",
       "0   130;194;145;135;167;3;120;170;61;112;126;143;1...   \n",
       "1              9;6;5;11;1;0;0;0;2;2;1;0;4;1;1;1;2;0;4   \n",
       "2                                                  NA   \n",
       "3   2;1;2;1;2;2;1;0;2;2;2;1;1;1;1;8;2;1;1;2;2;5;2;...   \n",
       "4   2;3;7;7;4;2;1;13;2;5;3;4;16;4;1;0;6;1;4;6;9;11...   \n",
       "..                                                ...   \n",
       "66                                                 NA   \n",
       "67                                                 NA   \n",
       "68                                                 NA   \n",
       "69  0;0;1;1;0;0;0;1;1;0;0;1;1;0;0;0;0;0;0;0;0;0;0;...   \n",
       "70                                                 NA   \n",
       "\n",
       "                           REGION.SNP.OTHER.HAP.COUNT  REGION.READ.COUNT  \\\n",
       "0   1;0;4;0;0;0;1;0;0;0;0;0;0;0;0;0;0;0;0;1;0;1;0;...             300944   \n",
       "1               0;0;0;3;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0              11727   \n",
       "2                                                  NA              33410   \n",
       "3   0;0;0;0;0;0;0;0;0;0;0;0;0;1;0;0;0;0;0;0;0;0;0;...               3039   \n",
       "4   0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;1;0;0;...              17576   \n",
       "..                                                ...                ...   \n",
       "66                                                 NA               6670   \n",
       "67                                                 NA             169811   \n",
       "68                                                 NA             169811   \n",
       "69  0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;...               7009   \n",
       "70                                                 NA                769   \n",
       "\n",
       "    GENOMEWIDE.READ.COUNT  \n",
       "0               428586390  \n",
       "1               428586390  \n",
       "2               428586390  \n",
       "3               428586390  \n",
       "4               428586390  \n",
       "..                    ...  \n",
       "66              428586390  \n",
       "67              428586390  \n",
       "68              428586390  \n",
       "69              428586390  \n",
       "70              428586390  \n",
       "\n",
       "[71 rows x 17 columns]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create the bespoke Huvec point libraries.\n",
    "# These join together donors based upon shared snp genotypes in the reg activity.\n",
    "\n",
    "from new_skew_wasp_input import get_haplotypes_counts, get_other_haplotype_counts\n",
    "from io import StringIO\n",
    "\n",
    "# Run for HUVEC_D1, HUVEC_D2, HUVEC_D3\n",
    "donor = \"HUVEC_D3\"\n",
    "\n",
    "# Need to load from both VCF files. First file has phasing information.\n",
    "# Second file has precomputed counts information for all datasets including POINT-seq\n",
    "vcf = pysam.VariantFile(phased_vcf_file_format.format(donor))\n",
    "vcf_counts = pysam.VariantFile(bam_counts_format.format(donor=donor))\n",
    "\n",
    "# Compute total read counts.\n",
    "stats = pd.read_csv(\n",
    "    StringIO(pysam.idxstats(\n",
    "        point_bam_format.format(donor)\n",
    "    )),\n",
    "    sep=\"\\t\",\n",
    "    header=None\n",
    ")\n",
    "genome_wide_read_count = stats[2].sum()\n",
    "print(genome_wide_read_count)\n",
    "\n",
    "output = []\n",
    "atac_input = None\n",
    "for _, row in test_frame[test_frame.donor == donor].iterrows():\n",
    "    \n",
    "    gene = genes[genes.gene_name == row.gene].iloc[0]\n",
    "    \n",
    "    read_count = pysam.AlignmentFile(point_bam_format.format(donor)).count(gene.chrom, gene.start, gene.end)\n",
    "\n",
    "    # Load all variants from gene region\n",
    "    variants = list(vcf.fetch(gene.chrom, gene.start, gene.end))\n",
    "    variant_counts = list(vcf_counts.fetch(gene.chrom, gene.start, gene.end))\n",
    "    assert len(variants) == len(variant_counts)\n",
    "    \n",
    "    # Restrict to only simply mono-allelic heterozygous SNPs\n",
    "    variants = [v for v in variants if set(v.samples[donor][\"GT\"]) == {0, 1}]\n",
    "    variant_counts = [v for v in variant_counts if set(v.samples[donor][\"GT\"]) == {0, 1}]\n",
    "    \n",
    "    # Restrict only to SNPs in phase with activity SNP. Using PacBio phasesets\n",
    "    variants = [v for v in variants if v.samples[donor].get(\"PS\") == row[\"phaseset\"]]\n",
    "    variant_counts = [v for v in variant_counts if v.samples[donor].get(\"PS\") == row[\"phaseset\"]]\n",
    "\n",
    "    # Restrict to SNPs disregarding Indel counts wihtin Gene body\n",
    "    variants = [v for v in variants if len(v.alts) == 1 and len(v.alts[0]) == len(v.ref)]\n",
    "    variant_counts = [v for v in variant_counts if len(v.alts) == 1 and len(v.alts[0]) == len(v.ref)]\n",
    "    assert len(variants) == len(variant_counts)\n",
    "    \n",
    "    # Get counts from forward stranded bam.\n",
    "    experiment_key = \"PF\" if gene.strand == \"+\" else \"PR\"\n",
    "    \n",
    "    h1, h2 = get_haplotypes_counts(variant_counts, donor, experiment_key, True)\n",
    "    other_counts = get_other_haplotype_counts(variant_counts, experiment_key)\n",
    "    \n",
    "    print(row[\"TEST.SNP.HAPLOTYPE\"])\n",
    "    if row[\"TEST.SNP.HAPLOTYPE\"] == \"0|1\":\n",
    "        ref_counts = h1\n",
    "        alt_counts = h2\n",
    "    else:\n",
    "        ref_counts = h2\n",
    "        alt_counts = h1\n",
    "    \n",
    "    # Write entry line for WASP input\n",
    "    not_empty = len(variant_counts) > 0\n",
    "    v_pos = [v.pos for v in variant_counts]\n",
    "    adjusted_het_probs = [1.0 for v in variant_counts]\n",
    "    hets_phase_probs = [1.0 for v in variant_counts]\n",
    "    \n",
    "    output.append({\n",
    "        \"CHROM\": row[\"CHROM\"],\n",
    "        \"TEST.SNP.POS\": row[\"TEST.SNP.POS\"],\n",
    "        \"TEST.SNP.ID\": row[\"TEST.SNP.ID\"],\n",
    "        \"TEST.SNP.REF.ALLELE\": row[\"TEST.SNP.REF.ALLELE\"],\n",
    "        \"TEST.SNP.ALT.ALLELE\": row[\"TEST.SNP.ALT.ALLELE\"],\n",
    "        \"TEST.SNP.GENOTYPE\": row[\"TEST.SNP.GENOTYPE\"],\n",
    "        # Ideally this should include possibility not phased for clear understanding\n",
    "        \"TEST.SNP.HAPLOTYPE\": row[\"TEST.SNP.HAPLOTYPE\"],\n",
    "        \"REGION.START\": gene.start,\n",
    "        \"REGION.END\": gene.end,\n",
    "        \"REGION.SNP.POS\": \";\".join(list(map(str, v_pos))) if not_empty else \"NA\",\n",
    "\n",
    "        \"REGION.SNP.HET.PROB\": \";\".join(list(map(str, adjusted_het_probs))) if not_empty else \"NA\",\n",
    "        \"REGION.SNP.LINKAGE.PROB\": \";\".join(list(map(str, hets_phase_probs))) if not_empty else \"NA\",\n",
    "\n",
    "        \"REGION.SNP.REF.HAP.COUNT\": \";\".join(list(map(str, ref_counts))) if not_empty else \"NA\",\n",
    "        \"REGION.SNP.ALT.HAP.COUNT\": \";\".join(list(map(str, alt_counts))) if not_empty else \"NA\",\n",
    "        \"REGION.SNP.OTHER.HAP.COUNT\": \";\".join(list(map(str, other_counts))) if not_empty else \"NA\",\n",
    "\n",
    "        \"REGION.READ.COUNT\": read_count,\n",
    "        \"GENOMEWIDE.READ.COUNT\": genome_wide_read_count,\n",
    "    })\n",
    "\n",
    "output = pd.DataFrame(output)\n",
    "\n",
    "output.to_csv(bespoke_wasp_input_format.format(donor), sep=\" \", index=False)\n",
    "output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "14f18689",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "299b4fc7",
   "metadata": {},
   "source": [
    "## Analysis of results. Reformat for Matt's spreadsheet\n",
    "\n",
    "This second part of the notebook is to be run after wasp is run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "91c5458a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(71, 19)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TEST.SNP.CHROM</th>\n",
       "      <th>TEST.SNP.POS</th>\n",
       "      <th>TEST.SNP.ID</th>\n",
       "      <th>TEST.SNP.REF.ALLELE</th>\n",
       "      <th>TEST.SNP.ALT.ALLELE</th>\n",
       "      <th>REGION.START</th>\n",
       "      <th>REGION.END</th>\n",
       "      <th>LOGLIKE.NULL</th>\n",
       "      <th>LOGLIKE.ALT</th>\n",
       "      <th>CHISQ</th>\n",
       "      <th>...</th>\n",
       "      <th>TOTAL.AS.READ.COUNT</th>\n",
       "      <th>REF.AS.READ.COUNT</th>\n",
       "      <th>ALT.AS.READ.COUNT</th>\n",
       "      <th>TOTAL.READ.COUNT</th>\n",
       "      <th>REGION</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2232129</td>\n",
       "      <td>rs263533</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>2227388</td>\n",
       "      <td>2310213</td>\n",
       "      <td>-18459.27</td>\n",
       "      <td>-18407.60</td>\n",
       "      <td>103.337</td>\n",
       "      <td>...</td>\n",
       "      <td>26958</td>\n",
       "      <td>14619</td>\n",
       "      <td>12339</td>\n",
       "      <td>485295</td>\n",
       "      <td>chr1:2227388-2310213</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2227388</td>\n",
       "      <td>2310213</td>\n",
       "      <td>+</td>\n",
       "      <td>SKI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2232129</td>\n",
       "      <td>rs263533</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>2184461</td>\n",
       "      <td>2212720</td>\n",
       "      <td>-250.36</td>\n",
       "      <td>-248.84</td>\n",
       "      <td>3.037</td>\n",
       "      <td>...</td>\n",
       "      <td>361</td>\n",
       "      <td>162</td>\n",
       "      <td>199</td>\n",
       "      <td>18281</td>\n",
       "      <td>chr1:2184461-2212720</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2184461</td>\n",
       "      <td>2212720</td>\n",
       "      <td>-</td>\n",
       "      <td>FAAP20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>56411837</td>\n",
       "      <td>rs11206803</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>56494761</td>\n",
       "      <td>56645301</td>\n",
       "      <td>-101.83</td>\n",
       "      <td>-101.79</td>\n",
       "      <td>0.096</td>\n",
       "      <td>...</td>\n",
       "      <td>151</td>\n",
       "      <td>71</td>\n",
       "      <td>80</td>\n",
       "      <td>130707</td>\n",
       "      <td>chr1:56494761-56645301</td>\n",
       "      <td>chr1</td>\n",
       "      <td>56494761</td>\n",
       "      <td>56645301</td>\n",
       "      <td>-</td>\n",
       "      <td>PLPP3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>205141789</td>\n",
       "      <td>rs3902193</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>205227946</td>\n",
       "      <td>205285632</td>\n",
       "      <td>-131.92</td>\n",
       "      <td>-130.94</td>\n",
       "      <td>1.966</td>\n",
       "      <td>...</td>\n",
       "      <td>192</td>\n",
       "      <td>86</td>\n",
       "      <td>106</td>\n",
       "      <td>3418</td>\n",
       "      <td>chr1:205227946-205285632</td>\n",
       "      <td>chr1</td>\n",
       "      <td>205227946</td>\n",
       "      <td>205285632</td>\n",
       "      <td>+</td>\n",
       "      <td>TMCC2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>205141789</td>\n",
       "      <td>rs3902193</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>205142505</td>\n",
       "      <td>205211702</td>\n",
       "      <td>-1067.08</td>\n",
       "      <td>-1066.39</td>\n",
       "      <td>1.366</td>\n",
       "      <td>...</td>\n",
       "      <td>1537</td>\n",
       "      <td>744</td>\n",
       "      <td>793</td>\n",
       "      <td>32457</td>\n",
       "      <td>chr1:205142505-205211702</td>\n",
       "      <td>chr1</td>\n",
       "      <td>205142505</td>\n",
       "      <td>205211702</td>\n",
       "      <td>-</td>\n",
       "      <td>DSTYK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>chr6</td>\n",
       "      <td>108902728</td>\n",
       "      <td>rs11961585</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>108848416</td>\n",
       "      <td>108974476</td>\n",
       "      <td>-4.16</td>\n",
       "      <td>-3.83</td>\n",
       "      <td>0.654</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>14077</td>\n",
       "      <td>chr6:108848416-108974476</td>\n",
       "      <td>chr6</td>\n",
       "      <td>108848416</td>\n",
       "      <td>108974476</td>\n",
       "      <td>+</td>\n",
       "      <td>ARMC2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>chr8</td>\n",
       "      <td>133540375</td>\n",
       "      <td>rs59413766</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>133454848</td>\n",
       "      <td>133571926</td>\n",
       "      <td>-4142.14</td>\n",
       "      <td>-4139.65</td>\n",
       "      <td>4.978</td>\n",
       "      <td>...</td>\n",
       "      <td>5974</td>\n",
       "      <td>2870</td>\n",
       "      <td>3104</td>\n",
       "      <td>336809</td>\n",
       "      <td>chr8:133454848-133571926</td>\n",
       "      <td>chr8</td>\n",
       "      <td>133454848</td>\n",
       "      <td>133571926</td>\n",
       "      <td>-</td>\n",
       "      <td>ST3GAL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>chr8</td>\n",
       "      <td>133546137</td>\n",
       "      <td>rs16904940</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>133454848</td>\n",
       "      <td>133571926</td>\n",
       "      <td>-4142.14</td>\n",
       "      <td>-4139.65</td>\n",
       "      <td>4.978</td>\n",
       "      <td>...</td>\n",
       "      <td>5974</td>\n",
       "      <td>2870</td>\n",
       "      <td>3104</td>\n",
       "      <td>336809</td>\n",
       "      <td>chr8:133454848-133571926</td>\n",
       "      <td>chr8</td>\n",
       "      <td>133454848</td>\n",
       "      <td>133571926</td>\n",
       "      <td>-</td>\n",
       "      <td>ST3GAL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>chr9</td>\n",
       "      <td>129168215</td>\n",
       "      <td>rs10819474</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>129608884</td>\n",
       "      <td>129636135</td>\n",
       "      <td>-154.34</td>\n",
       "      <td>-154.02</td>\n",
       "      <td>0.647</td>\n",
       "      <td>...</td>\n",
       "      <td>224</td>\n",
       "      <td>121</td>\n",
       "      <td>103</td>\n",
       "      <td>11413</td>\n",
       "      <td>chr9:129608884-129636135</td>\n",
       "      <td>chr9</td>\n",
       "      <td>129608884</td>\n",
       "      <td>129636135</td>\n",
       "      <td>+</td>\n",
       "      <td>NTMT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>chr9</td>\n",
       "      <td>129168215</td>\n",
       "      <td>rs10819474</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>129175552</td>\n",
       "      <td>129178261</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1084</td>\n",
       "      <td>chr9:129175552-129178261</td>\n",
       "      <td>chr9</td>\n",
       "      <td>129175552</td>\n",
       "      <td>129178261</td>\n",
       "      <td>-</td>\n",
       "      <td>IER5L</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>71 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   TEST.SNP.CHROM  TEST.SNP.POS TEST.SNP.ID TEST.SNP.REF.ALLELE  \\\n",
       "0            chr1       2232129    rs263533                   C   \n",
       "1            chr1       2232129    rs263533                   C   \n",
       "2            chr1      56411837  rs11206803                   C   \n",
       "3            chr1     205141789   rs3902193                   C   \n",
       "4            chr1     205141789   rs3902193                   C   \n",
       "..            ...           ...         ...                 ...   \n",
       "66           chr6     108902728  rs11961585                   G   \n",
       "67           chr8     133540375  rs59413766                   C   \n",
       "68           chr8     133546137  rs16904940                   C   \n",
       "69           chr9     129168215  rs10819474                   T   \n",
       "70           chr9     129168215  rs10819474                   T   \n",
       "\n",
       "   TEST.SNP.ALT.ALLELE  REGION.START  REGION.END  LOGLIKE.NULL  LOGLIKE.ALT  \\\n",
       "0                    T       2227388     2310213     -18459.27    -18407.60   \n",
       "1                    T       2184461     2212720       -250.36      -248.84   \n",
       "2                    T      56494761    56645301       -101.83      -101.79   \n",
       "3                    A     205227946   205285632       -131.92      -130.94   \n",
       "4                    A     205142505   205211702      -1067.08     -1066.39   \n",
       "..                 ...           ...         ...           ...          ...   \n",
       "66                   A     108848416   108974476         -4.16        -3.83   \n",
       "67                   A     133454848   133571926      -4142.14     -4139.65   \n",
       "68                   T     133454848   133571926      -4142.14     -4139.65   \n",
       "69                   C     129608884   129636135       -154.34      -154.02   \n",
       "70                   C     129175552   129178261          0.00         0.00   \n",
       "\n",
       "      CHISQ  ...  TOTAL.AS.READ.COUNT  REF.AS.READ.COUNT  ALT.AS.READ.COUNT  \\\n",
       "0   103.337  ...                26958              14619              12339   \n",
       "1     3.037  ...                  361                162                199   \n",
       "2     0.096  ...                  151                 71                 80   \n",
       "3     1.966  ...                  192                 86                106   \n",
       "4     1.366  ...                 1537                744                793   \n",
       "..      ...  ...                  ...                ...                ...   \n",
       "66    0.654  ...                    6                  2                  4   \n",
       "67    4.978  ...                 5974               2870               3104   \n",
       "68    4.978  ...                 5974               2870               3104   \n",
       "69    0.647  ...                  224                121                103   \n",
       "70    0.000  ...                    0                  0                  0   \n",
       "\n",
       "    TOTAL.READ.COUNT                    REGION  chrom      start        end  \\\n",
       "0             485295      chr1:2227388-2310213   chr1    2227388    2310213   \n",
       "1              18281      chr1:2184461-2212720   chr1    2184461    2212720   \n",
       "2             130707    chr1:56494761-56645301   chr1   56494761   56645301   \n",
       "3               3418  chr1:205227946-205285632   chr1  205227946  205285632   \n",
       "4              32457  chr1:205142505-205211702   chr1  205142505  205211702   \n",
       "..               ...                       ...    ...        ...        ...   \n",
       "66             14077  chr6:108848416-108974476   chr6  108848416  108974476   \n",
       "67            336809  chr8:133454848-133571926   chr8  133454848  133571926   \n",
       "68            336809  chr8:133454848-133571926   chr8  133454848  133571926   \n",
       "69             11413  chr9:129608884-129636135   chr9  129608884  129636135   \n",
       "70              1084  chr9:129175552-129178261   chr9  129175552  129178261   \n",
       "\n",
       "   strand gene_name  \n",
       "0       +       SKI  \n",
       "1       -    FAAP20  \n",
       "2       -     PLPP3  \n",
       "3       +     TMCC2  \n",
       "4       -     DSTYK  \n",
       "..    ...       ...  \n",
       "66      +     ARMC2  \n",
       "67      -   ST3GAL1  \n",
       "68      -   ST3GAL1  \n",
       "69      +     NTMT1  \n",
       "70      -     IER5L  \n",
       "\n",
       "[71 rows x 24 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes[\"REGION\"] = genes[\"chrom\"] + \":\" + genes[\"start\"].astype(str) + \"-\" + genes[\"end\"].astype(str)\n",
    "\n",
    "bespoke_wasp = None\n",
    "bespoke_wasp = pd.read_csv(\"/project/Wellcome_Discovery/esanders/01_huvec_skew/08_bespoke_wasp/cht_results.2.txt\", sep=\"\\t\")\n",
    "\n",
    "bespoke_wasp[\"REGION\"] = bespoke_wasp[\"TEST.SNP.CHROM\"] + \":\" + bespoke_wasp[\"REGION.START\"].astype(str) + \"-\" + bespoke_wasp[\"REGION.END\"].astype(str)\n",
    "\n",
    "print(bespoke_wasp.shape)\n",
    "\n",
    "bespoke_wasp = bespoke_wasp.merge(genes, on=\"REGION\")\n",
    "\n",
    "bespoke_wasp.to_csv(output_summary_filepath, sep=\"\\t\", index=False)\n",
    "bespoke_wasp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07c88963",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b4489be",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
